File: Codeforces/492B.cpp
#include <bits/stdc++.h>
#define IOS ios_base::sync_with_stdio(false); cin.tie(NULL);
#define loop(i,n) for(int i=0;i<n;i++)
#define for1(i,n) for(int i=1;i<=n;i++)
typedef long long ll;
using namespace std;
bool isPossible(vector<int>&arr,int mid,int k){
     ll left = 0;
     ll right = 0;
     bool flag = false;
     for(int i=0;i<arr.size();i++){
          if(arr[i]-mid<=right){
               right = arr[i]+mid;
          } else {
               flag = true;
               break;
          } if (right>=k*1e9) break;
     }
     return flag;
}
void solve(){ 
     ll n,k;
     cin>>n>>k;
     vector<ll> arr(n);
     loop(i,n)cin>>arr[i];
     sort(arr.begin(),arr.end());
     ll d=0;
	for (int i = 1; i < n; ++i)  // find the max gap
	{
		d = max(d,arr[i]-arr[i-1]);
	}
	long long int x = max(arr[0],k-arr[n-1]); 
	double y = d/2.0;
	if(y >  x)
          printf("%.9f",y);
	else
		cout<<x<<endl;
     
}
int main() {
int t=1;
cin>>t;
while(t--){
 solve(); 
}
     return 0;
}

File: Codeforces/492B.py
def check(arr,d,m):
      x = arr[0]
      if(x > d):
            return False
      for i in range(1,len(arr)):
         x = (arr[i]-arr[i-1])/2.0
         if(x > d):
            return False
      if(m-arr[-1]>x): return False
      return True

def solve(): 
    n,m=map(int,input().split())
    
    arr = [int(x) for x in input().split()]
    arr.sort()

    low = 0
    high = m
    ans = max(arr)
    while low <= high :
        mid = low + (high-low)/2.0
        # print(mid)
        if (check(arr,mid,m)):
            ans = mid
            high = mid-0.001
        else:
            low = mid + 0.001
    # print(high , low)
    print("%.1f"%ans)


 
if __name__ == '__main__' :
  t = 1
  t= int(input())
  while(t>0):
     solve() 
     t-=1